xarray/tests/duckarrays/base/strategies.py

```diff
@@ -0,0 +1,199 @@
+from typing import Any
+
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+
+import xarray as xr
+from xarray.core.utils import is_dict_like
+
+from . import utils
+
+all_dtypes = (
+    npst.integer_dtypes()
+    | npst.unsigned_integer_dtypes()
+    | npst.floating_dtypes()
+    | npst.complex_number_dtypes()
+)
+
+
+def numpy_array(shape, dtypes=None):
+    if dtypes is None:
+        dtypes = all_dtypes
+
+    def elements(dtype):
+        max_value = 100
+        min_value = 0 if dtype.kind == "u" else -max_value
+
+        return npst.from_dtype(
+            dtype, allow_infinity=False, min_value=min_value, max_value=max_value
+        )
+
+    return dtypes.flatmap(
+        lambda dtype: npst.arrays(dtype=dtype, shape=shape, elements=elements(dtype))
+    )
+
+
+def dimension_sizes(min_dims, max_dims, min_size, max_size):
+    sizes = st.lists(
+        elements=st.tuples(st.text(min_size=1), st.integers(min_size, max_size)),
+        min_size=min_dims,
+        max_size=max_dims,
+        unique_by=lambda x: x[0],
+    )
+    return sizes
+
+
+# Is there a way to do this in general?
+# Could make a Protocol...
+T_DuckArray = Any
+
+
+@st.composite
+def duckarray(
+    draw,
+    create_data,
+    *,
+    sizes=None,
+    min_size=1,
+    max_size=3,
+    min_dims=1,
+    max_dims=3,
+    dtypes=None,
+) -> st.SearchStrategy[T_DuckArray]:
```

Comment:
I tried to add a new strategy for creating a duck array of varying shape/dtype, but ideally I would like to refactor the other strategies to depend on this one.